### PR TITLE
SLA Policy (user side)

### PR DESCRIPTION
**Pull request by rafalpospiech**
obraz
image

### CODE REVIEW

File: app/Utils/ServiceContracts.php [hunk 1]
```diff
@@ -85,6 +85,21 @@ private static function getDefaultBusinessHours(): array
...
 	}
 
+	/**
+	 * Get all business hours.
+	 *
+	 * @return array
+	 */
+	public static function getAllBusinessHours(): array
+	{
+		if (\App\Cache::has('UtilsServiceContracts::getDefaultBusinessHours', '')) {
```
Comment (mariuszkrzaczkowski) [inline app/Utils/ServiceContracts.php]:
```suggestion
		if (\App\Cache::has('UtilsServiceContracts::getAllBusinessHours', '')) {
```

File: app/Utils/ServiceContracts.php [hunk 2]
```diff
@@ -85,6 +85,21 @@ private static function getDefaultBusinessHours(): array
...
+			return \App\Cache::get('UtilsServiceContracts::getDefaultBusinessHours', '');
```
Comment (mariuszkrzaczkowski) [inline app/Utils/ServiceContracts.php]:
```suggestion
			return \App\Cache::get('UtilsServiceContracts::getAllBusinessHours', '');
```

File: app/Utils/ServiceContracts.php [hunk 3]
```diff
@@ -85,6 +85,21 @@ private static function getDefaultBusinessHours(): array
...
+		}
+		$rows = (new \App\Db\Query())->from('s_#__business_hours')->all(\App\Db::getInstance('admin'));
+		\App\Cache::save('UtilsServiceContracts::getDefaultBusinessHours', '', $rows);
```
Comment (mariuszkrzaczkowski) [inline app/Utils/ServiceContracts.php]:
```suggestion
		\App\Cache::save('UtilsServiceContracts::getAllBusinessHours', '', $rows);
```

File: modules/ServiceContracts/actions/PolicyDeleteAjax.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+			throw new \App\Exceptions\NoPermitted('LBL_PERMISSION_DENIED', 406);
+		}
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function process(App\Request $request)
+	{
+		\App\Db::getInstance()->createCommand()->delete('u_#__servicecontracts_sla_policy', ['id' => $request->getInteger('record')])->execute();
```
Comment (TomaszQr) [inline modules/ServiceContracts/actions/PolicyDeleteAjax.php]: MVC?

File: modules/ServiceContracts/actions/PolicySaveAjax.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+			} else {
+				$data['conditions'] = '';
+			}
+			$data['business_hours'] = implode(',', $request->getArray('business_hours', 'Integer')[$rowIndex]);
+			$data['reaction_time'] = $request->getArray('reaction_time', 'TimePeriod')[$rowIndex];
+			$data['idle_time'] = $request->getArray('idle_time', 'TimePeriod')[$rowIndex];
+			$data['resolve_time'] = $request->getArray('resolve_time', 'TimePeriod')[$rowIndex];
+			$data['crmid'] = $crmId;
+			$data['tabid'] = \App\Module::getModuleId($request->getByType('targetModule', 'Alnum'));
+			$db->createCommand()->insert('u_#__servicecontracts_sla_policy', $data)->execute();
```
Comment (TomaszQr) [inline modules/ServiceContracts/actions/PolicySaveAjax.php]: MVC?

File: modules/ServiceContracts/actions/PolicyTemplatesAjax.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function process(App\Request $request)
+	{
+		$recordModels = Settings_SlaPolicy_Record_Model::getForModule($request->getByType('targetModule', 'Alnum'));
+		$recordId = $request->getInteger('recordId');
+		$current = (new \App\Db\Query())->select('sla_policy_id')->from('u_#__servicecontracts_sla_policy')->where(['crmid' => $recordId])->scalar();
```
Comment (TomaszQr) [inline modules/ServiceContracts/actions/PolicyTemplatesAjax.php]: MVC?

**Pull request merged**